### PR TITLE
[GPU] gather nd shape agnostic kernel implementation

### PR DESCRIPTION
**Pull request by wilson-seok**
### Details:
 - gather nd shape agnostic kernel implementation

### Tickets:
 - 107778

### CODE REVIEW

File: src/plugins/intel_gpu/tests/unit/test_cases/gather_nd_gpu_test.cpp [hunk 1]
```diff
@@ -780,3 +781,184 @@ TEST(gather_nd_gpu_fp16, export_import) {
...
+    auto input2 = engine.allocate_memory(layout{ov::PartialShape(in2_shape), data_types::f16, format::bfzyx}); // Indexes
+
+    //auto input1 = engine.allocate_memory({ data_types::f16, format::bfyx, { 3, 3, 3, 2 } }); // data
+    //auto input2 = engine.allocate_memory({ data_types::f16, format::bfyx, { 3, 2, 3, 2 } }); // indices
+
+    //const int indices_rank = 5;
+    //const int batch_dims = 4;
+    //auto input0 = engine.allocate_memory({ data_types::f16, format::bfzyx, { 2, 3, 2, 2, 3 } }); // data
+    //auto input1 = engine.allocate_memory({ data_types::f16, format::bfzyx, { 2, 3, 1, 2, 3 } }); // indices
+    // expected output dim: v5{36}, v8{2,3,2,3}
```
Comment (andrew-k-park) [inline src/plugins/intel_gpu/tests/unit/test_cases/gather_nd_gpu_test.cpp:891]: Please remove unnecessary code

File: src/plugins/intel_gpu/src/graph/impls/ocl/gather_nd.cpp [hunk 2]
```diff
@@ -35,22 +36,45 @@ struct gather_nd_impl : typed_primitive_impl_ocl<gather_nd> {
...
+    };
+
+    auto static_formats = {
+        format::bfyx,
+        format::bfzyx,
+        format::bfwzyx
+    };
+
+    implementation_map<gather_nd>::add(impl_types::ocl,
+                                    shape_types::static_shape,
```
Comment (andrew-k-park) [inline src/plugins/intel_gpu/src/graph/impls/ocl/gather_nd.cpp:62]: nit: please correct for indentation

File: src/plugins/intel_gpu/src/graph/impls/ocl/gather_nd.cpp [hunk 3]
```diff
@@ -35,22 +36,45 @@ struct gather_nd_impl : typed_primitive_impl_ocl<gather_nd> {
...
+                                    static_formats);
+
+    auto dyn_formats = {
+        format::bfyx,
+        format::bfzyx,
+        format::bfwzyx
+    };
+
+    implementation_map<gather_nd>::add(impl_types::ocl,
+                                    shape_types::dynamic_shape,
```
Comment (andrew-k-park) [inline src/plugins/intel_gpu/src/graph/impls/ocl/gather_nd.cpp:74]: nit: same as above

File: src/plugins/intel_gpu/src/kernel_selector/kernels/gather/gather_nd_kernel_ref.cpp [hunk 4]
```diff
@@ -155,9 +156,32 @@ bool GatherNDKernelRef::Validate(const Params& p, const optional_params& o) cons
...
+                return false;
+            }
+        }
+    }
+
+    if (params.outputs[0].is_dynamic()) {
+        auto supported_tensor_layout = [](const DataTensor& t) -> bool {
+            if (t.GetLayout() == DataLayout::bfyx ||
+                t.GetLayout() == DataLayout::bfzyx ||
+                t.GetLayout() == DataLayout::bfwzyx) {
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/gather/gather_nd_kernel_ref.cpp:171]: Do we need this check here? It is already checked in impelementation_map

File: src/plugins/intel_gpu/src/kernel_selector/kernels/gather/gather_nd_kernel_ref.cpp [hunk 5]
```diff
@@ -177,13 +201,35 @@ KernelsData GatherNDKernelRef::GetKernelsData(const Params& params, const option
     KernelData kd = KernelData::Default<gather_nd_params>(params);
     gather_nd_params& newParams = *static_cast<gather_nd_params*>(kd.params.get());
 
-    auto dispatchData = SetDefault(newParams, options);
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/gather/gather_nd_kernel_ref.cpp:180]: why this changed is needed?

File: src/plugins/intel_gpu/src/graph/impls/ocl/gather_nd.cpp [hunk 6]
```diff
@@ -23,6 +23,7 @@ struct gather_nd_impl : typed_primitive_impl_ocl<gather_nd> {
         return make_unique<gather_nd_impl>(*this);
     }
 
+public:
```
Comment (e-ddykim) [inline src/plugins/intel_gpu/src/graph/impls/ocl/gather_nd.cpp:26]: Why did you add this? It is already "public".

File: src/plugins/intel_gpu/src/kernel_selector/kernels/gather/gather_nd_kernel_ref.cpp (hunk 5 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/gather/gather_nd_kernel_ref.cpp:180]: This is not critical, I just followed gather case(https://github.com/openvinotoolkit/openvino/pull/13987).

File: src/plugins/intel_gpu/src/graph/impls/ocl/gather_nd.cpp (hunk 2 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/impls/ocl/gather_nd.cpp]: Fixed. Thanks!

File: src/plugins/intel_gpu/src/graph/impls/ocl/gather_nd.cpp (hunk 3 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/impls/ocl/gather_nd.cpp]: Fixed. Thanks!

File: src/plugins/intel_gpu/src/kernel_selector/kernels/gather/gather_nd_kernel_ref.cpp (hunk 4 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/gather/gather_nd_kernel_ref.cpp:171]: Agreed and Fixed. Thanks!

File: src/plugins/intel_gpu/tests/unit/test_cases/gather_nd_gpu_test.cpp (hunk 1 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/tests/unit/test_cases/gather_nd_gpu_test.cpp]: Fixed. Thanks!

File: src/plugins/intel_gpu/src/graph/impls/ocl/gather_nd.cpp (hunk 6 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/impls/ocl/gather_nd.cpp]: FIxed. Thanks!

File: src/plugins/intel_gpu/src/kernel_selector/kernels/gather/gather_nd_kernel_ref.cpp (hunk 5 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/gather/gather_nd_kernel_ref.cpp:180]: Not only gather but also several kernels merged recently have same change. I think it is better to sync with them.

**Review by yeonbok — APPROVED** [review]
